### PR TITLE
support generating new image with separate blob for prefetched files

### PR DESCRIPTION
**Pull request by Maxing1997**
## Relevant Issue https://github.com/dragonflyoss/nydus/issues/1589
The objective is to generate a new image, which references a new blob included prefetch file chunks. This will accelerate an  startup performance due to high-priority pulling of prefetch blob.

## Details
Implement optimize subcommand of nydusify to generate a new image, which references a new blob included prefetch file chunks.

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [x] New feature (non-breaking change which adds functionality)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: contrib/nydusify/pkg/optimizer/optimizer.go [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+			}
+		} else {
+			return errors.Wrap(err, "pull source image")
+		}
+	}
+	logrus.Infof("pulled source image, elapsed: %s", time.Since(start))
+	return nil
+}
+
+// Optimize covert and push a new optimized nydus image
```
Comment (imeoer) [inline contrib/nydusify/pkg/optimizer/optimizer.go:235]: `covert` -> `converts`

File: contrib/nydusify/pkg/optimizer/optimizer.go [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+func Optimize(ctx context.Context, opt Opt) error {
+	ctx = namespaces.WithNamespace(ctx, "nydusify")
+
+	sourceRemote, err := provider.DefaultRemote(opt.Source, opt.SourceInsecure)
+	if err != nil {
+		return errors.Wrap(err, "Init source image parser")
+	}
+	sourceParser, err := parser.New(sourceRemote, runtime.GOARCH)
+	if sourceParser == nil {
```
Comment (imeoer) [inline contrib/nydusify/pkg/optimizer/optimizer.go:244]: `if err != nil {`

File: contrib/nydusify/pkg/optimizer/optimizer.go [hunk 3]
```diff
@@ -0,0 +1,536 @@
...
+				return nil, errors.Wrap(err, "push blob")
+			}
+		} else {
+			return nil, errors.Wrap(err, "push blob")
+		}
+	}
+	return &blobDesc, nil
+}
+
+func pushCompoundBootstrap(ctx context.Context, opt Opt, buildInfo BuildInfo) (*bootstrapInfo, error) {
```
Comment (imeoer) [inline contrib/nydusify/pkg/optimizer/optimizer.go:361]: `pushCompoundBootstrap` -> `pushNewBootstrap` ?

File: contrib/nydusify/pkg/optimizer/optimizer.go [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+	tarDigester := digest.SHA256.Digester()
+	if _, err := io.Copy(io.MultiWriter(bootstrapTar, tarDigester.Hash()), rc); err != nil {
+		return nil, errors.Wrap(err, "get tar digest")
+	}
+	bootstrapDiffID := tarDigester.Digest()
+
+	bootstrapTarRa, err := os.Open(bootstrapTarPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "open bootstrap tar file")
+	}
```
Comment (imeoer) [inline contrib/nydusify/pkg/optimizer/optimizer.go:404]: defer bootstrapTarRa.Close()

**Comment by imeoer** [comment]
Thanks!

File: contrib/nydusify/cmd/nydusify.go [hunk 5]
```diff
@@ -1160,6 +1162,97 @@ func main() {
 				return copier.Copy(context.Background(), opt)
 			},
 		},
+		{
+			Name:  "optimize",
+			Usage: "Convert and push a new optimized nydus image",
```
Comment (imeoer) [inline contrib/nydusify/cmd/nydusify.go:1167]: Optimize a source nydus image and push to the target

File: contrib/nydusify/pkg/optimizer/builder.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	OutputBootstrapPath string
+	OutputJSONPath      string
+	Timeout             *time.Duration
+}
+
+type outputJSON struct {
+	Blobs []string `json:"blobs"`
+}
+
+func Build(option BuildOption) (string, error) {
```
Comment (imeoer) [inline contrib/nydusify/pkg/optimizer/builder.go:35]: `func Optimize`

File: contrib/nydusify/pkg/optimizer/builder.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	outputBytes, err := os.ReadFile(outputJSONPath)
+	if err != nil {
+		return "", errors.Wrapf(err, "read file %s", outputJSONPath)
+	}
+	var output outputJSON
+	err = json.Unmarshal(outputBytes, &output)
+	if err != nil {
+		return "", errors.Wrapf(err, "unmarshal output json file %s", outputJSONPath)
+	}
+	BlobID := output.Blobs[len(output.Blobs)-1]
```
Comment (imeoer) [inline contrib/nydusify/pkg/optimizer/builder.go:83]: `BlobID` -> `blobID`

File: storage/src/device.rs [hunk 8]
```diff
@@ -79,6 +79,8 @@ bitflags! {
         const _V5_NO_EXT_BLOB_TABLE = 0x8000_0000;
         /// Blob is generated with chunkdict.
         const IS_CHUNKDICT_GENERATED = 0x0000_0200;
+        /// Blob is generated with separated prefetch files.
+        const IS_SEPARATED_WITH_PREFETCH_FILES = 0x0001_0000;
```
Comment (imeoer) [inline storage/src/device.rs:83]: How about `0x0000_0400`?

File: contrib/nydusify/pkg/optimizer/builder.go (hunk 6 above)
Comment (Maxing1997) [reply contrib/nydusify/pkg/optimizer/builder.go:35]: this conflicts `Optimize` declaration in optimizer.go

**Review by imeoer — APPROVED** [review]
Thanks, LGTM

**Pull request merged**